src/trace/snapshotter.ts

```diff
@@ -0,0 +1,525 @@
+  const escaped = { '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', '\'': '&#39;' };
+
+  const escapeAttribute = (s: string): string => {
+    return s.replace(/[&<>"']/g, char => (escaped as any)[char]);
+  };
+  const escapeText = (s: string): string => {
+    return s.replace(/[&<]/g, char => (escaped as any)[char]);
+  };
+
+  const frames = new Map<string, Element>();
+  const chunks = new Map<string, string>();
+  const resources = new Set<string>();
+  let counter = 0;
+
+  const nextId = (): string => {
+    return guid + (++counter);
+  };
+
+  const resolve = (base: string, url: string): string => {
+    if (url === '')
+      return '';
+    try {
+      return new URL(url, base).href;
+    } catch (e) {
+      return url;
+    }
+  };
+
+  const addResource = (base: string, url: string): string => {
+    if (url.startsWith('javascript:'))
+      return '';
+    if (url.startsWith('data:') || !url)
+      return url;
+    resources.add(resolve(base, url));
+    return url;
+  };
+
+  const processSrcSet = (base: string, srcset: string): string => {
+    return srcset.split(',').map(src => {
+      src = src.trim();
+      const spaceIndex = src.lastIndexOf(' ');
+      if (spaceIndex === -1)
+        return addResource(base, src);
+      return addResource(base, src.substring(0, spaceIndex).trim()) + src.substring(spaceIndex);
+    }).join(',');
+  };
+
+  const processStyle = (base: string, style: string) => {
+    // TODO: this is wrong.
+    for (const match of style.matchAll(/(url\s*\(['"]?)([^)'"]*)(['"]?\))/g))
+      addResource(base, match[2]);
+  };
+
+  // TODO: handle CSSOM.
+  // TODO: handle fonts loaded with FontFaceSet.
+
+  const visit = (node: Node, builder: string[]) => {
+    if (node.nodeType === Node.DOCUMENT_TYPE_NODE) {
+      const docType = node as DocumentType;
+      builder.push(`<!DOCTYPE ${docType.name}>`);
+      return;
+    }
+
+    if (node.nodeType === Node.TEXT_NODE) {
+      builder.push(node.parentElement!.nodeName === 'STYLE' ? (node.nodeValue || '') : escapeText(node.nodeValue || ''));
+      return;
+    }
+
+    if (node.nodeType !== Node.ELEMENT_NODE &&
+        node.nodeType !== Node.DOCUMENT_NODE &&
+        node.nodeType !== Node.DOCUMENT_FRAGMENT_NODE)
+      return;
+
+    if (node.nodeName === 'SCRIPT' || node.nodeName === 'BASE')
+      return;
+
+    if (removeNoScript && node.nodeName === 'NOSCRIPT')
+      return;
+
+    if (node.nodeName === 'STYLE')
+      processStyle(node.baseURI, node.textContent || '');
+
+    if (node.nodeType === Node.ELEMENT_NODE) {
+      const element = node as Element;
+      builder.push('<');
+      builder.push(element.nodeName);
+      for (let i = 0; i < element.attributes.length; i++) {
+        const name = element.attributes[i].name;
+        let value = element.attributes[i].value;
+        if (name.toLowerCase() === 'value' && (element.nodeName === 'INPUT' || element.nodeName === 'TEXTAREA'))
```

Comment:
Extract it? Looks expensive to call toLowerCase on every attribute of every node 7 times?